paimon-hive/paimon-hive-connector-common/src/main/java/org/apache/paimon/hive/HiveSchema.java

```diff
@@ -86,33 +95,67 @@ public static HiveSchema extract(@Nullable Configuration configuration, Properti
-                            "column.name.delimite", String.valueOf(SerDeUtils.COMMA));
-            List<String> names = Arrays.asList(columnNames.split(columnNameDelimiter));
-
-            String columnTypes = properties.getProperty(serdeConstants.LIST_COLUMN_TYPES);
-            List<TypeInfo> typeInfos = TypeInfoUtils.getTypeInfosFromTypeString(columnTypes);
-
-            if (names.size() > 0 && typeInfos.size() > 0) {
-                checkSchemaMatched(names, typeInfos, tableSchema);
+        Optional<TableSchema> tableSchemaOptional = TableUtils.schema(catalogContext);
+
+        String columnProperty = properties.getProperty(serdeConstants.LIST_COLUMNS);
+        // Create hive external table with empty ddl
+        if (StringUtils.isEmpty(columnProperty)) {
+            if (!tableSchemaOptional.isPresent()) {
+                throw new IllegalArgumentException(
+                        "Schema file not found in location "
+                                + location
+                                + ". Please create table first.");
             }
+            // Paimon external table can read schema from the specified location
+            return new HiveSchema(tableSchemaOptional.get());
         }
 
+        // Create hive external table with ddl
+        String columnNameDelimiter =
+                properties.getProperty(
+                        // serdeConstants.COLUMN_NAME_DELIMITER is not defined in earlier Hive
+                        // versions, so we use a constant string instead
+                        "column.name.delimite", String.valueOf(SerDeUtils.COMMA));
+        List<String> columnNames = Arrays.asList(columnProperty.split(columnNameDelimiter));
+        String columnTypes = properties.getProperty(serdeConstants.LIST_COLUMN_TYPES);
+        List<TypeInfo> typeInfos = TypeInfoUtils.getTypeInfosFromTypeString(columnTypes);
+        List<String> comments =
+                Lists.newArrayList(
+                        Splitter.on('\0').split(properties.getProperty("columns.comments")));
+        // Both Paimon table schema and Hive table schema exist
+        if (tableSchemaOptional.isPresent() && columnNames.size() > 0 && typeInfos.size() > 0) {
+            LOG.debug(
+                    "Extract schema with exists DDL and exists paimon table, table location:[{}].",
+                    location);
+            checkSchemaMatched(columnNames, typeInfos, tableSchemaOptional.get());
+            comments = schemaComments(tableSchemaOptional.get());
+        }
+
+        int highestFieldId = -1;
+        List<DataField> columns = new ArrayList<>();
+        for (int i = 0; i < columnNames.size(); i++) {
+            columns.add(
+                    new DataField(
+                            ++highestFieldId,
+                            columnNames.get(i),
+                            typeInfoToLogicalType(typeInfos.get(i)),
+                            comments.get(i)));
+        }
+        TableSchema tableSchema =
```

Comment:
Can you do a refactor to `HiveSchema`?
It does not need `TableSchema`, only rowType is OK.

Reply:
done.